cmd/fs-v1-multipart.go

```diff
@@ -601,52 +755,42 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
 	}
 	fsMeta.Meta["md5Sum"] = s3MD5
 
-	fsMetaPath = path.Join(bucketMetaPrefix, bucket, object, fsMetaJSONFile)
-	// Write the metadata to a temp file and rename it to the actual location.
-	if err = writeFSMetadata(fs.storage, minioMetaBucket, fsMetaPath, fsMeta); err != nil {
+	// Write all the set metadata.
+	if _, err = fsMeta.WriteTo(metaFile); err != nil {
 		return "", toObjectErr(err, bucket, object)
 	}
 
-	// Cleanup all the parts if everything else has been safely committed.
-	if err = cleanupUploadedParts(bucket, object, uploadID, fs.storage); err != nil {
-		return "", toObjectErr(err, bucket, object)
-	}
-
-	// Hold the lock so that two parallel
-	// complete-multipart-uploads do not leave a stale
-	// uploads.json behind.
+	// Hold the lock so that two parallel complete-multipart-uploads
+	// do not leave a stale uploads.json behind.
 	objectMPartPathLock := globalNSMutex.NewNSLock(minioMetaMultipartBucket,
 		pathJoin(bucket, object))
 	objectMPartPathLock.Lock()
 	defer objectMPartPathLock.Unlock()
 
-	// remove entry from uploads.json
-	if err := fs.removeUploadID(bucket, object, uploadID); err != nil {
-		return "", toObjectErr(err, minioMetaMultipartBucket, path.Join(bucket, object))
+	// Disallow any parallel abort or complete multipart operations.
+	rwlk, err := fs.rwPool.Create(pathJoin(fs.fsPath, minioMetaMultipartBucket, bucket, object, uploadsJSONFile))
+	if err != nil {
+		return "", toObjectErr(traceError(err), bucket, object)
 	}
+	defer rwlk.Close()
 
-	// Return md5sum.
-	return s3MD5, nil
-}
+	// Close lock held on bucket/object/uploadid/fs.json
+	fs.rwPool.Close(fsMetaPathMultipart)
```

Comment:
double Close() will cause problems, i.e it can close the fd even when its not supposed to.